text/0000-image-labels.md

```diff
@@ -0,0 +1,69 @@
+# Meta
+[meta]: #meta
+- Name: Image Labels
+- Start Date: 2020-03-24
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Buildpacks should be able to expose arbitrary label metadata on images.  This content should be exposed by enhancing buildpack artifacts.
+
+# Motivation
+[motivation]: #motivation
+
+While buildpacks create OCI images, the standard does not expose all elements that OCI images can support to buildpack authors.  One of these elements, [image labels][l], is regularly requested by users in order to contribute both [standard][s] and arbitrary metadata to their generated images.  Once delivered, this functionality would let buildpack authors contribute labels to the images their buildpacks create.
+
+[l]: https://docs.docker.com/engine/reference/builder/#label
+[s]: https://github.com/opencontainers/image-spec/blob/master/annotations.md#pre-defined-annotation-keys
+
+# What it is
+[what-it-is]: #what-it-is
+
+Buildpacks must be able to contribute arbitrary key-value pairs that will be placed as labels on images.  This can be added as a non-breaking change to `launch.toml` as a peer to `processes` and `slices`. It should be comprised of a collection of items, each with `key` and `value` entries.
+
+```toml
+[[labels]]
+key = "<label key 1>"
+value = "<label value 1>"
+
+[[labels]]
+key = "<label key 2>"
+value = "<label value 2>"
```

Comment:
Is there a specific distinction as to why these would be added to the `launch.toml` vs a directory structure similar to that of environment variables?

Reply:
Creating a directory structure assumes that the relationship will only ever be key-value and can never be extended beyond that.  The `processes` collection, with it's late addition of `direct` is a good example where the flexibility of a collection of TOML objects is useful.

Reply:
With that in mind, isn't labels key/value. Are we anticipating that `labels` will get additions?

From my limited-knowledge perspective, it seems odd that what I consider global attributes (env and labels) are set via two _very_ different mechanism. Would we consider moving _env_ to `launch.toml` otherwise?

Reply:
Labels are key-value today, but it's easy to envision some kind of conditional or late reconciliation, etc. possibility in the future and there's no need to close that functionality off today by choosing an inflexible representation.

`/env` is a different case since one of the most important target platforms (Kubernetes) already has an opinion on how to represent environment variables (value-per-file).

Reply:
I suddenly don’t like the name `launch.toml` if it now contains image labels, processes, and slices. But maybe that is just me.

Reply:
@jabrown85 we've talked about changing the name of `launch.toml` before. I think we should consider changing it at the same time we consider changing the keys in the `<layer>.toml` (`build`, `cache`, `launch`)

Reply:
To @jromero 's suggestion about adding labels with `<layer>/labels/KEY` files instead of in `launch.toml`, I think there a practical dealbreaker to consider here in addition to the format concern:

* Adding labels in a directory structure would embed the information in the layer itself, so a reused launch layer that is not restored to disk on a rebuild would lose its labels. The fact that `exporter` needs to read the labels makes it materially different from launch time env vars, which are read by the launcher.

Another tradeoff to consider:
* Putting labels in `launch.toml` makes them a per-buildpack concern rather than a per-layer concern. This means that buildpacks must regenerate labels on every build. **I feel like this is the correct call**. If for any reason we wanted a reused layer to bring its labels with it, they would need to be in `<layer.toml>`.